### PR TITLE
add snapshot testing

### PR DESCRIPTION
**Pull request by pngwn**
- adds snapshot testing capability 
- updates some test deps
- cleans up tests to use some nicer selectors
- expanded the generate config script to generate configs for snapshots

The snapshots aren't currently running in CI because I'm not sure what the best way is to generate them. Ideally the CI machine would generate the snapshots and save them + update them when needed but that is moderately complex + i don't really CI pushing commits to active branches when a human might also be working on them. For now the ability to generate snapshots is more of a local thing that people can use when working on a PR, of limited value but a start.

I'm not super happy with how we are generating the necessary demos because it means we need a dedicated file, which a specific name, per demo which is fine when we have a bunch of tests but slightly annoying for snapshots. I'll see what alternative options we have.

Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: # (issue)

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

File: ui/packages/app/test/blocks_kinematics.spec.ts [hunk 1]
```diff
@@ -30,6 +30,8 @@ test("renders the correct elements", async ({ page }) => {
 	await mock_api(page, [[25, 45]]);
 	await page.goto("http://localhost:3000");
 
-	const button = await page.locator("button");
-	await Promise.all([button.click(), page.waitForResponse("**/api/predict/")]);
+	await Promise.all([
+		page.click("button"),
+		page.waitForResponse("**/api/predict/")
+	]);
```
Comment (pngwn) [inline ui/packages/app/test/blocks_kinematics.spec.ts:36]: Not sure that this test is really doing anything.

File: ui/packages/app/test/blocks_page_load.spec.ts [hunk 2]
```diff
@@ -30,13 +30,11 @@ test("renders the correct elements", async ({ page }) => {
...
 
-	const textbox = await page.locator("label:has-text('Name')");
+	const textbox = await page.getByLabel("Name");
 
 	await textbox.fill("Frank");
-	await expect(await page.inputValue("label:has-text('Name')")).toEqual(
-		"Frank"
-	);
-	await expect(await page.inputValue("label:has-text('Output')")).toEqual(
+	await expect(await textbox.inputValue()).toEqual("Frank");
```
Comment (pngwn) [inline ui/packages/app/test/blocks_page_load.spec.ts]: use `toHaveValue()`

File: ui/packages/app/test/outbreak_forecast.spec.ts [hunk 3]
```diff
@@ -30,17 +30,13 @@ test("matplotlib", async ({ page }) => {
...
-		.selectOption("Matplotlib");
-	await page
-		.locator("text=Month JanuaryFebruaryMarchAprilMay >> select")
-		.selectOption("January");
-	await page.locator('label:has-text("Social Distancing?")').click();
 
-	const submit_button = await page.locator("text=Submit");
+	await page.getByLabel("Plot Type").selectOption("Matplotlib");
+	await page.getByLabel("Month").selectOption("January");
+	await page.getByLabel("Social Distancing?").click();
```
Comment (pngwn) [inline ui/packages/app/test/outbreak_forecast.spec.ts]: `check()`

File: ui/packages/app/test/outbreak_forecast.spec.ts [hunk 4]
```diff
@@ -60,17 +56,13 @@ test("plotly", async ({ page }) => {
...
-		.selectOption("Plotly");
-	await page
-		.locator("text=Month JanuaryFebruaryMarchAprilMay >> select")
-		.selectOption("January");
-	await page.locator('label:has-text("Social Distancing?")').click();
 
-	const submit_button = await page.locator("text=Submit");
+	await page.getByLabel("Plot Type").selectOption("Plotly");
+	await page.getByLabel("Month").selectOption("January");
+	await page.getByLabel("Social Distancing?").click();
```
Comment (pngwn) [inline ui/packages/app/test/outbreak_forecast.spec.ts]: `check()`

**Comment by dawoodkhan82** [comment]
I'm getting an error for screenshot comparison failed, when trying to run the screenshot test twice. Am I doing something wrong? @pngwn

**Comment by dawoodkhan82** [comment]
Thanks for cleaning up the tests with nicer selectors!

**Comment by pngwn** [comment]
@dawoodkhan82 Yeah, the new selectors were only added recently but they are pretty nice!

Regarding the screenshots, that is expected right now. I've gitignored them for now. Currently the flow would be to generate the screenshots once and then use those generated screenshots as a reference for any changes you make. 

The first time you run the screenshot tests they will fail (and generate the screen shots for you), subsequent runs (without UI changes) should pass.

So right now it is a bit clunky but the flow would be:

- run tests to generate screenshots
- make changes
- test against screenshots and either fix the regression or update the screenshot (there is a separate command to update screenshots i think)

**Comment by dawoodkhan82** [comment]
> 

Sorry to clarify, after running twice it fails for me. Heres the screenshots it saved:

Expected:
blocks-layout-1-expected

Actual:
blocks-layout-1-actual

**Comment by dawoodkhan82** [comment]
> @dawoodkhan82 Yeah, the new selectors were only added recently but they are pretty nice!
> 
> Regarding the screenshots, that is expected right now. I've gitignored them for now. Currently the flow would be to generate the screenshots once and then use those generated screenshots as a reference for any changes you make.
> 
> The first time you run the screenshot tests they will fail (and generate the screen shots for you), subsequent runs (without UI changes) should pass.
> 
> So right now it is a bit clunky but the flow would be:
> 
> * run tests to generate screenshots
> * make changes
> * test against screenshots and either fix the regression or update the screenshot (there is a separate command to update screenshots i think)

Ok I see what I did, the first time I ran it I didn't have the frontend running on localhost:3000 so it took an empty screenshot to compare it too. Works now though

**Review by dawoodkhan82 — APPROVED** [review]

File: .github/workflows/ui.yml [hunk 5]
```diff
@@ -54,11 +54,11 @@ jobs:
           cache: pnpm
           cache-dependency-path: ui/pnpm-lock.yaml
       - run: pip install -r ../requirements.txt
+      - run: pip install -r ../demo/outbreak_forecast/requirements.txt
```
[the remark this answers is not in the document]
Comment (pngwn) [reply .github/workflows/ui.yml:57]: If you know of a way we can remove this lmk, and we can fix in another PR!

**Pull request merged**